src/OmniSharp.Roslyn.CSharp/Workers/Diagnostics/CSharpDiagnosticService.cs

```diff
@@ -1,152 +1,222 @@
-                    }
-                }
+                var currentWork = _workQueue
+                    .Where(x => x.Value.modified.AddMilliseconds(_throttlingMs) < DateTime.UtcNow)
+                    .OrderByDescending(x => x.Value.modified) // If you currently edit project X you want it will be highest priority and contains always latest possible analysis.
+                    .Take(2) // Limit mount of work executed by once. This is needed on large solution...
+                    .ToList();
 
-                if (!_queueRunning && !_openDocuments.IsEmpty)
+                foreach (var workKey in currentWork.Select(x => x.Key))
                 {
-                    this.ProcessQueue();
+                    _workQueue.TryRemove(workKey, out _);
                 }
+
+                return currentWork.Select(x => (x.Value.project, x.Value.workReadySource));
             }
         }
 
-        private void ProcessQueue()
+        public async Task<IEnumerable<(string projectName, Diagnostic diagnostic)>> GetCurrentDiagnosticResult(IEnumerable<ProjectId> projectIds)
+        {
+            await WaitForInitialStartupWorkIfAny();
+
+            var pendingWork = WaitForPendingWorkIfNeededAndGetIt(projectIds);
+
+            await Task.WhenAll(pendingWork);
+
+            return _results
+                .Where(x => projectIds.Any(pid => pid == x.Key))
+                .SelectMany(x => x.Value.diagnostics, (k, v) => ((k.Value.name, v)));
+        }
+
+        public void QueueForAnalysis(IEnumerable<Project> projects)
         {
-            lock (_lock)
+            foreach (var project in projects)
             {
-                _queueRunning = true;
+                _workQueue.AddOrUpdate(project.Id,
+                    (modified: DateTime.UtcNow, project: project, new CancellationTokenSource()),
+                    (_, oldValue) => (modified: DateTime.UtcNow, project: project, oldValue.workReadySource));
             }
+        }
 
-            Task.Factory.StartNew(async () =>
-            {
-                await Task.Delay(100);
-                await Dequeue();
+        private IEnumerable<Task> WaitForPendingWorkIfNeededAndGetIt(IEnumerable<ProjectId> projectIds)
+        {
+            return _workQueue
+                .Where(x => projectIds.Any(pid => pid == x.Key))
+                .Select(x => Task.Delay(10 * 1000, x.Value.workReadySource.Token)
+                    .ContinueWith(task => LogTimeouts(task, x.Key.ToString())));
+        }
 
-                if (_openDocuments.IsEmpty)
-                {
-                    lock (_lock)
-                    {
-                        _queueRunning = false;
-                    }
-                }
-                else
-                {
-                    this.ProcessQueue();
-                }
-            });
+        private Task WaitForInitialStartupWorkIfAny()
+        {
+            return Task.Delay(10 * 1000, _initializationQueueDoneSource.Token)
```

Comment:
not sure what this wait is trying to do.

Reply:
VScode atleast seems to make initial request for all diagnostics available in beginning and it doesn't re-request without user change to another project or something. For this reason api doesn't return before workspace is ready and initial analysis is done. If this isn't made first editor shows 0 diagnostic and suddenly it had all some time later when it made new request for all diagnostics. I don't know is there good way to do this with current api design.

Reply:
I will add comments about those cases to code, they are kind of hacks.